tools/revpi-set-mac

```diff
@@ -36,22 +36,32 @@ case $dev_type in
         size=256
         eeprom_file='/usr/share/revpi/lan9514.bin'
         ;;
+    ks8851)
+        magic='0x8851'
+        # offset needs to be 7 to 2 (first to last byte)
+        offset_delta=-8
+        offset_multiplier=-1
+        ;;
     *)
         1>&2 echo "Unsupported device type: '$dev_type'"
         exit 1
         ;;
 esac
 
 # write golden image to EEPROM
-if [ -f "${eeprom_file}" ]; then
-    ethtool -E "${iface}" magic "${magic}" offset 0 length "${size}" < "${eeprom_file}"
-else
-    echo 1>&2 "ERROR: Golden image not found: '${eeprom_file}"
-    exit 1
+if [ -n "${eeprom_file:-}" ]; then
+    if [ -f "${eeprom_file}" ]; then
+        ethtool -E "${iface}" magic "${magic}" offset 0 length "${size}" < "${eeprom_file}"
+    else
+        echo 1>&2 "ERROR: Golden image not found: '${eeprom_file}"
+        exit 1
+    fi
 fi
 
 # write MAC Address
 for idx in {1..6}
 do
-    ethtool -E "${iface}" magic "${magic}" offset "${idx}" length 1 value 0x"${BASH_REMATCH[idx]}"
+    offset=$(((idx + ${offset_delta:-0})*${offset_multiplier:-1}))
```

Comment:
This is pretty genius but honestly I fear it will be difficult for someone else to read and understand what's going on here. (Heck, *I* will have difficulty remembering it 1 year later.)

My recommendation would be to instead [reverse](https://stackoverflow.com/questions/11461625/reverse-the-order-of-characters-in-a-string) the MAC address in `ks8851-set-mac`, then pass that to `revpi-set-mac`. I.e., keep the special-handling required for the ks8851 in its script instead of cluttering up common code.

An alternative approach would be to create a library which can be `source`d by `ks8851-set-mac` and which provides a function to reverse the MAC address.

Reply:
I can see your point and agree that this needs better explanation for others (even the future me). But I don't like the approach of reversing the mac address as user input. This breaks the functionality of  `revpi-set-mac` as single script for all devices (the other scripts can be seen as convenience / legacy tooling), when the user needs to remember the bytes order depending on the used ethernet controller. How about reversing the mac in the `ks8851` section of the case?

Reply:
With a bit more thinking I came to the conclusion, that reversing the MAC alone won't do the trick as the offset is also different. That's the initial reason why I came up with the formula above.

Reply:
> How about reversing the mac in the `ks8851` section of the case?

I guess that would also work.

Reply:
> With a bit more thinking I came to the conclusion, that reversing the MAC alone won't do the trick as the offset is also different. That's the initial reason why I came up with the formula above.

Actually there's no reason to write the MAC address bytewise. Why not write it all at once and set the offset in the dev_type section for the ks8851 further up in the script?

Reply:
Honestly I just copied the parts from the former ks8851-set-mac script and mixed-in the offset stuff. The idea was to have "one script to rule them all". I've added some comments and also stated the defaults to my offset calculation. Please have a look if we an agree on this generic solution. I'm not sure if writing the MAC for all devices will work en-bloc (eg. byte wise was stated in microchips application notes). In order to have a solution which needs less as possible variants.

Reply:
"One script to rule them all"...  Best phrase of the day

Reply:
> The idea was to have "one script to rule them all".

One hard-learned lesson of the kernel community is to avoid midlayers and use libraries instead.

Recommended reading: <a href="https://lwn.net/Articles/336262/">LWN article</a> by Neil Brown and <a href="https://blog.ffwll.ch/2016/12/midlayers-once-more-with-feeling.html">Blog posting</a> by Daniel Vetter with additional links and background info.

Basically the lesson is that trying to put everything and the kitchen sink in a "fat" midlayer is a bad idea. It quickly becomes unwieldy, unmaintainable and difficult to read. Your midlayer in this case is the `revpi-set-mac` script. You're trying to handle every possible quirk of every Ethernet chip ever used in a RevPi in a single, "fat" script and the per-chip scripts such as `ks8851-set-mac` merely become "thin" wrappers. From a software engineering perspective, it is better to provide a library of functions which perform specific tasks such as: Validate if this string contains a MAC address, write this golden image, etc. Then you'd call these library functions from the per-chip scripts. Hence my suggestion to move reversal of the MAC address to `ks8851-set-mac`. That way you'd at least encapsulate that quirk of a single chip in its chip-specific script, instead of inflating the "midlayer" script.

You can ignore the hard-learned lessons of the kernel community, but at your own peril. ;)

Reply:
Awesome, thanks a lot. Makes sense to keep that in a separate pull request to avoid scope creep of this one.